src/private.rs

```diff
@@ -0,0 +1,256 @@
+            } else {
+                Poll::Pending
+            }
+        }
+    }
+
+    // pub struct MultiNotificationInner<const N: usize> {
+    //     waker: MultiWakerRegistration<N>,
+    //     notified: bool,
+    // }
+
+    // pub struct MultiNotification<const N: usize> {
+    //     cs: IsrCriticalSection,
+    //     inner: UnsafeCell<MultiNotificationInner<N>>,
+    // }
+
+    // impl<const N: usize> MultiNotification<N> {
+    //     pub const fn new() -> Self {
+    //         Self {
+    //             cs: IsrCriticalSection::new(),
+    //             inner: UnsafeCell::new(MultiNotificationInner {
+    //                 waker: MultiWakerRegistration::new(),
+    //                 notified: false,
+    //             }),
+    //         }
+    //     }
+
+    //     pub fn notify(&self) -> bool {
+    //         let _cs = self.cs.enter();
+
+    //         let inner = unsafe { self.inner.get().as_mut() }.unwrap();
+
+    //         inner.notified = true;
+
+    //         inner.waker.wake()
+    //     }
+
+    //     pub fn wait(&self) -> impl Future<Output = ()> + '_ {
+    //         core::future::poll_fn(move |cx| self.poll_wait(cx))
+    //     }
+
+    //     pub fn poll_wait(&self, cx: &Context<'_>) -> Poll<()> {
+    //         let _cs = self.cs.enter();
+
+    //         let inner = unsafe { self.inner.get().as_mut() }.unwrap();
+
+    //         inner.waker.register(cx.waker()).unwrap();
+
+    //         if inner.notified {
+    //             Poll::Ready(())
+    //         } else {
+    //             Poll::Pending
+    //         }
+    //     }
+    // }
+
+    // unsafe impl<const N: usize> Send for MultiNotification<N> {}
+    // unsafe impl<const N: usize> Sync for MultiNotification<N> {}
+}
+
+pub mod completion {
+    use core::{
+        future::Future,
+        pin::Pin,
+        task::{Context, Poll},
+    };
+
+    pub struct Completion<F, D>
+    where
+        D: FnMut(bool),
+    {
+        fut: F,
+        destr: D,
+        completed: bool,
+    }
+
+    impl<F, D> Completion<F, D>
+    where
+        D: FnMut(bool),
+    {
+        pub const fn new(fut: F, destr: D) -> Self {
+            Self {
+                fut,
+                destr,
+                completed: false,
+            }
+        }
+    }
+
+    impl<F, D> Future for Completion<F, D>
+    where
+        F: Future,
+        D: FnMut(bool),
+    {
+        type Output = F::Output;
+
+        fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+            let this = unsafe { self.get_unchecked_mut() };
+
+            let fut = unsafe { Pin::new_unchecked(&mut this.fut) };
+
+            let poll = fut.poll(cx);
+
+            this.completed |= matches!(&poll, Poll::Ready(_));
+
+            poll
+        }
+    }
+
+    impl<F, D> Drop for Completion<F, D>
+    where
+        D: FnMut(bool),
+    {
+        fn drop(&mut self) {
+            (self.destr)(self.completed);
```

Comment:
Any reason not to do something like this instead?
```suggestion
            if !self.completed {
                (self.destr)();
            }
```

Reply:
Yes: more control to the API user, as their callback will be called regardless.

Reply:
API user is just `esp-idf-hal` right? Do any of the callbacks actually run code when `completed` is true?

Reply:
For now - no. Tomorrow - I have no idea. In any case this looks like nit-picking to me. Let's just leave it as -is.

Reply:
The semantics just seem strange to me since the API user can easily handle the completed case themselves.
```
Completion::new(..., ...).await;

// code for completed case.
```

But you're right, at the end of the day I'm just nit-picking.
(I'd resolve the conversation if GitHub let me)